server/tests/api/test_rounds.py

```diff
@@ -209,6 +209,56 @@ def test_rounds_complete_audit(
     compare_json(rounds, expected_rounds)
 
 
+def test_rounds_round_2_required_if_all_blanks(
+    client: FlaskClient, election_id: str, contest_ids: List[str], round_1_id: str,
+):
+    run_audit_round_all_blanks(round_1_id, contest_ids[0], contest_ids)
+    rv = client.get(f"/api/election/{election_id}/round")
+    rounds = json.loads(rv.data)["rounds"]
+    assert len(rounds) == 1
+    assert_is_date(rounds[0]["endedAt"])
+    assert not rounds[0]["isAuditComplete"]
+
+
+def test_rounds_end_logic_unaffected_by_invalid_write_ins_1(
+    client: FlaskClient, election_id: str, contest_ids: List[str], round_1_id: str,
+):
+    run_audit_round(
+        round_1_id, contest_ids[0], contest_ids, 0.7, invalid_write_in_ratio=1
+    )
+    rv = client.get(f"/api/election/{election_id}/round")
+    rounds = json.loads(rv.data)["rounds"]
+    assert len(rounds) == 1
+    assert_is_date(rounds[0]["endedAt"])
+    assert rounds[0]["isAuditComplete"]
+
+
+def test_rounds_end_logic_unaffected_by_invalid_write_ins_2(
+    client: FlaskClient, election_id: str, contest_ids: List[str], round_1_id: str,
+):
+    run_audit_round(
+        round_1_id, contest_ids[0], contest_ids, 0.5, invalid_write_in_ratio=1
+    )
+    rv = client.get(f"/api/election/{election_id}/round")
+    rounds = json.loads(rv.data)["rounds"]
+    assert len(rounds) == 1
+    assert_is_date(rounds[0]["endedAt"])
+    assert not rounds[0]["isAuditComplete"]
+
+
+def test_rounds_end_logic_unaffected_by_invalid_write_ins_3(
+    client: FlaskClient, election_id: str, contest_ids: List[str], round_1_id: str,
+):
+    run_audit_round_all_blanks(
+        round_1_id, contest_ids[0], contest_ids, invalid_write_in_ratio=1
+    )
+    rv = client.get(f"/api/election/{election_id}/round")
+    rounds = json.loads(rv.data)["rounds"]
+    assert len(rounds) == 1
+    assert_is_date(rounds[0]["endedAt"])
+    assert not rounds[0]["isAuditComplete"]
```

Comment:
Thanks for adding these. This is great.

Hate to do this to ya but I think we also need tests for ballot comparison. You might be able to modify an existing test or two to use invalid write-ins and just verify that the outcome didn't change.